### PR TITLE
Bug/msg session reconnect

### PR DESCRIPTION
**Pull request by dpwspoon**
**Issue #:**

**Description of changes:**

Fix issue with recent change to getMessagingSessionEndpoint resolve on connect where it would not reconnect if getMessagingSessionEndpoint failed.

**Testing:**

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*

Start the messaging demo app in browser.  Create app instance user.  Connect.  Disconnect wifi, reconnect, and see websocket get establish

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
yes

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
yes

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
yes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/messagingsession/DefaultMessagingSession.ts [hunk 1]
```diff
@@ -191,16 +193,21 @@ export default class DefaultMessagingSession implements MessagingSession {
     }
   }
 
+  private retryConnection(): boolean {
+    return this.reconnectController.retryWithBackoff(async () => {
+      try {
+        await this.startConnecting(true);
+      } catch (e) {
+        // getMessagingSessionEndpoint can fail, so need to retry here
+        this.retryConnection();
```
Comment (dpwspoon) [inline src/messagingsession/DefaultMessagingSession.ts:202]: should this be schedule next tick or something so call stack doesn't get large?

File: src/messagingsession/DefaultMessagingSession.ts [hunk 2]
```diff
@@ -191,16 +193,21 @@ export default class DefaultMessagingSession implements MessagingSession {
     }
   }
 
+  private retryConnection(): boolean {
+    return this.reconnectController.retryWithBackoff(async () => {
+      try {
+        await this.startConnecting(true);
+      } catch (e) {
```
Comment (devalevenkatesh) [inline src/messagingsession/DefaultMessagingSession.ts:200]: So this behavior is a new thing? Should not the existing code run into this as well?

File: src/messagingsession/DefaultMessagingSession.ts (hunk 1 above)
Comment (devalevenkatesh) [inline src/messagingsession/DefaultMessagingSession.ts:202]: Should there be a threshold on how long one can retry connecting? I think the threshold is handled by `retryWithBackoff` already so it should execute only when the `retryConnection` returns `true` right?

I think this should be return

File: src/messagingsession/DefaultMessagingSession.ts [hunk 3]
```diff
@@ -107,6 +107,15 @@ export default class DefaultMessagingSession implements MessagingSession {
...
+      }
+      if (this.reconnectController.hasStartedConnectionAttempt()) {
+        this.reconnectController.startedConnectionAttempt(false);
+      } else {
+        this.reconnectController.startedConnectionAttempt(true);
+      }
+
       // reconnect needs to re-resolve endpoint url, which will also refresh credentials on client if they are expired.
       let endpointUrl = !reconnecting ? this.configuration.endpointUrl : undefined;
       if (endpointUrl === undefined) {
```
Comment (devalevenkatesh) [inline src/messagingsession/DefaultMessagingSession.ts:121]: I think you also have to handle the error here, sorry I missed that there is no `catch` here:
https://github.com/aws/amazon-chime-sdk-js/pull/2179#discussion_r860332917

Comment (devalevenkatesh) [reply src/messagingsession/DefaultMessagingSession.ts:121]: The `getMessagingSessionEndpoint()` can error out here as well.

File: src/messagingsession/DefaultMessagingSession.ts (hunk 1 above)
Comment (devalevenkatesh) [inline src/messagingsession/DefaultMessagingSession.ts:202]: I think `this.retryConnection()` needs to return a boolean value to check whether to trigger a reconnection or not right?

Something like:
```
private retryConnection(): boolean {
  try {
   if (this.reconnectController.retryWithBackoff(async () => await this.startConnecting(true))) {
     return true;
   } else {
     return false;
   }
  } catch (error) {
    this.logger.error(`${error}`);
    return retryConnection();
  }
}
```

File: src/messagingsession/DefaultMessagingSession.ts (hunk 3 above)
Comment (dpwspoon) [reply src/messagingsession/DefaultMessagingSession.ts:121]: There is a finally block around this whole thing to reset the boolean.  Not sure what if anything we should do with the caught exception.  Thoughts?

File: src/messagingsession/DefaultMessagingSession.ts [hunk 4]
```diff
@@ -126,6 +115,30 @@ export default class DefaultMessagingSession implements MessagingSession {
...
+      // reconnect needs to re-resolve endpoint url, which will also refresh credentials on client if they are expired.
+      let endpointUrl = !reconnecting ? this.configuration.endpointUrl : undefined;
+      if (endpointUrl === undefined) {
+        try {
+          const endpoint = await this.configuration.chimeClient
+            .getMessagingSessionEndpoint()
+            .promise();
+          endpointUrl = endpoint.Endpoint.Url;
+        } catch (e) {
+          const closeEvent = new CloseEvent('close', {
```
Comment (devalevenkatesh) [inline src/messagingsession/DefaultMessagingSession.ts:128]: Could you print the error here before creating the CloseEvent?
```
this.logger.error(${error});
```

File: src/messagingsession/DefaultMessagingSession.ts [hunk 5]
```diff
@@ -191,16 +204,18 @@ export default class DefaultMessagingSession implements MessagingSession {
...
-      !this.isClosing &&
-      this.canReconnect(event.code) &&
-      this.reconnectController.retryWithBackoff(async () => {
-        await this.startConnecting(true);
-      }, null)
-    ) {
+    if (event.code !== 4999) {
+      this.webSocket.destroy();
+    }
+    if (!this.isClosing && this.canReconnect(event.code) && this.retryConnection()) {
```
Comment (devalevenkatesh) [inline src/messagingsession/DefaultMessagingSession.ts:218]: Here, `canReconnect` does not handle `4999` and will return `false`, so, should it now handle this as we are throwing a custom `CloseEvent` with `4999` code and expect a retry and a reconnect?

Comment (dpwspoon) [reply src/messagingsession/DefaultMessagingSession.ts:218]: `canReconnect` does handle 4999 cause `closeCode > 4000 ` other than the ones special case returns true.

File: src/messagingsession/DefaultMessagingSession.ts (hunk 4 above)
Comment (dpwspoon) [reply src/messagingsession/DefaultMessagingSession.ts:128]: This is already logged when handled in the closeEventHandler so don't think duplicate logging is needed here: `  this.logger.info(`WebSocket close: ${event.code} ${event.reason}`);`

Comment (devalevenkatesh) [reply src/messagingsession/DefaultMessagingSession.ts:128]: Yeah, but the reason is logged, not sure if `error` can be something different hence suggested.

File: src/messagingsession/DefaultMessagingSession.ts (hunk 5 above)
Comment (devalevenkatesh) [reply src/messagingsession/DefaultMessagingSession.ts:218]: Ah, my bad, missed the `>`, thanks for correcting.

File: src/messagingsession/DefaultMessagingSession.ts (hunk 4 above)
Comment (devalevenkatesh) [reply src/messagingsession/DefaultMessagingSession.ts:128]: Also, good solution I think.

**Review by devalevenkatesh — APPROVED** [review]

**Pull request merged**